halo2-ecc/src/fields/native_fp.rs

```diff
@@ -0,0 +1,213 @@
+use super::{BigPrimeField, FieldChip, Selectable};
+use halo2_base::gates::RangeChip;
+use halo2_base::{gates::GateInstructions, utils::modulus, AssignedValue, Context};
+use num_bigint::BigUint;
+use std::marker::PhantomData;
+
+// native field chip which implements FieldChip, use GateInstructions for basic arithmetic operations
+#[derive(Clone, Debug)]
+pub struct NativeFieldChip<'range, F: BigPrimeField> {
+    pub range: &'range RangeChip<F>,
+    pub native_modulus: BigUint,
+    _marker: PhantomData<F>,
+}
+
+impl<'range, F: BigPrimeField> NativeFieldChip<'range, F> {
+    pub fn new(range: &'range RangeChip<F>) -> Self {
+        let native_modulus = modulus::<F>();
+        Self { range, native_modulus, _marker: PhantomData }
+    }
+}
+
+impl<'range, F: BigPrimeField> FieldChip<F> for NativeFieldChip<'range, F> {
+    const PRIME_FIELD_NUM_BITS: u32 = F::NUM_BITS;
+    type UnsafeFieldPoint = AssignedValue<F>;
+    type FieldPoint = AssignedValue<F>;
+    type ReducedFieldPoint = AssignedValue<F>;
+    type FieldType = F;
+    type RangeChip = RangeChip<F>;
+
+    fn native_modulus(&self) -> &BigUint {
+        &self.native_modulus
+    }
+    fn range(&self) -> &'range Self::RangeChip {
+        self.range
+    }
+    fn limb_bits(&self) -> usize {
+        F::CAPACITY as usize
+    }
+
+    fn get_assigned_value(&self, x: &AssignedValue<F>) -> F {
+        *x.value()
+    }
+
+    fn load_private(&self, ctx: &mut Context<F>, a: F) -> AssignedValue<F> {
+        ctx.load_witness(a)
+    }
+
+    fn load_constant(&self, ctx: &mut Context<F>, a: F) -> AssignedValue<F> {
+        ctx.load_constant(a)
+    }
+
+    // signed overflow BigInt functions
+    fn add_no_carry(
+        &self,
+        ctx: &mut Context<F>,
+        a: impl Into<AssignedValue<F>>,
+        b: impl Into<AssignedValue<F>>,
+    ) -> AssignedValue<F> {
+        self.gate().add(ctx, a.into(), b.into())
+    }
+
+    fn add_constant_no_carry(
+        &self,
+        ctx: &mut Context<F>,
+        a: impl Into<AssignedValue<F>>,
+        c: F,
+    ) -> AssignedValue<F> {
+        let c = self.load_constant(ctx, c);
+        self.add_no_carry(ctx, a, c)
```

Comment:
you can do `self.gate().add(ctx, a.into(), Constant(c))` to avoid the extra assignment in `load_constant`